src/sidebar/services/service-url.js

```diff
@@ -1,71 +1,34 @@
-import * as urlUtil from '../util/url';
-
 /**
- * A function that returns an absolute URL given a link name and params, by
- * expanding named URL templates received from the annotation service's API.
- *
- * The links object from the API is a map of link names to URL templates:
+ * Service for fetching the data needed to render URLs that point to the H
+ * service.
  *
- * {
- *   signup: "http://localhost:5000/signup",
- *   user: "http://localhost:5000/u/:user",
- *   ...
- * }
+ * The H API has an `/api/links` endpoint that returns a map of link name to
+ * URL template for URLs that point to the H API. This service fetches that
+ * data and persists it in the store.
  *
- * Given a link name (e.g. 'user') and params (e.g. {user: 'bob'}) return
- * an absolute URL by expanding the named URL template from the API with the
- * given params (e.g. "http://localhost:5000/u/bob").
+ * To use a link within a UI component, use `store.getLink(name, params)`.
  *
- * Before the links object has been received from the API this function
- * always returns empty strings as the URLs. After the links object has been
- * received from the API this function starts returning the real URLs.
- *
- * @callback ServiceUrlGetter
- * @param {string} linkName - The name of the link to expand
- * @param {object} [params] - The params with which to expand the link
- * @returns {string} The expanded absolute URL, or an empty string if the
- *                   links haven't been received from the API yet
- * @throws {Error} If the links have been received from the API but the given
- *                 linkName is unknown
- * @throws {Error} If one or more of the params given isn't used in the URL
- *                 template
- */
-
-/**
- * @param {import('../store').SidebarStore} store
- * @param {import('./api-routes').APIRoutesService} apiRoutes
- * @return {ServiceUrlGetter}
  * @inject
  */
-export default function serviceUrl(store, apiRoutes) {
-  apiRoutes
-    .links()
-    .then(store.updateLinks)
-    .catch(function (error) {
-      console.warn('The links API request was rejected: ' + error.message);
-    });
-
-  return function (linkName, params) {
-    const links = store.getState().links;
-
-    if (links === null) {
-      return '';
+export class ServiceURLService {
+  /**
+   * @param {import('./api-routes').APIRoutesService} apiRoutes
+   * @param {import('../store').SidebarStore} store
+   */
+  constructor(apiRoutes, store) {
+    this._apiRoutes = apiRoutes;
+    this._store = store;
+  }
+
+  /**
+   * Fetch URL templates for links from the API and persist them in the store.
+   */
+  async init() {
+    try {
+      const links = await this._apiRoutes.links();
+      this._store.updateLinks(links);
+    } catch (error) {
+      console.warn('Failed to fetch Hypothesis links: ' + error.message);
```

Comment:
Yesterday I said that it would be a good idea to retry but I left it out of scope. This morning I realized that `APIRouteService#links` already handles retrying the fetch, so we'll only enter this `catch` block if several retries failed.